### PR TITLE
feature(enhancement): TikTok_Ads - adding event mapping feature

### PR DESCRIPTION
**Pull request by rohithkaza**
## Description of the change

> Adding Event Mapping Fearture

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/tiktok_ads/transform.js [hunk 1]
```diff
@@ -33,18 +35,111 @@ function checkIfValidPhoneNumber(str) {
...
 
-  if (eventNameMapping[event] === undefined) {
+  const standardEventsMap = getHashFromArrayWithDuplicate(eventsToStandard);
+  const trimmedEvent = event.toLowerCase().trim();
+
+  if (
+    eventNameMapping[event] === undefined &&
+    standardEventsMap &&
+    !standardEventsMap[trimmedEvent]
+  ) {
```
Comment (ItsSudip) [inline v0/destinations/tiktok_ads/transform.js:124]:
```suggestion
  if (
    eventNameMapping[event] === undefined && !standardEventsMap[trimmedEvent]
  ) {
```
`getHashFromArrayWithDuplicate` function always returns one object so no need to check that.

File: v0/destinations/tiktok_ads/transform.js [hunk 2]
```diff
@@ -110,6 +204,7 @@ const trackResponseBuilder = async (message, { Config }) => {
   response.method = defaultPostRequestConfig.requestMethod;
   response.endpoint = TRACK_ENDPOINT;
   response.body.JSON = removeUndefinedAndNullValues(payload);
+
   return response;
```
Comment (ItsSudip) [inline v0/destinations/tiktok_ads/transform.js:208]: Line no 143 to 208 is redundant. We can remove that by calling `getTrackResponse` function.

File: v0/util/index.js [hunk 3]
```diff
@@ -120,6 +120,39 @@ const getHashFromArray = (
   return hashMap;
 };
 
+/**
+ * Format the destination.Config.dynamicMap arrays to hashMap
+ * Convert an array map to hashmap(value as an array)
```
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/util/index.js]: Done

**Review by ItsSudip — DISMISSED** [review]

File: v0/destinations/tiktok_ads/transform.js [hunk 4]
```diff
@@ -110,9 +100,48 @@ const trackResponseBuilder = async (message, { Config }) => {
...
 };
 
+const trackResponseBuilder = async (message, { Config }) => {
+  const pixel_code = Config.pixelCode;
```
Comment (utsabc) [inline v0/destinations/tiktok_ads/transform.js:108]: `pixel_code` is unused

File: v0/destinations/tiktok_ads/transform.js [hunk 5]
```diff
@@ -110,9 +100,48 @@ const trackResponseBuilder = async (message, { Config }) => {
...
+  const eventsToStandard = Config.eventsToStandard;
+
+  let event = get(message, "event");
+  event = event ? event.trim().toLowerCase() : event;
+  if (!event) {
+    throw new CustomError("Event name is required", 400);
+  }
```
Comment (utsabc) [inline v0/destinations/tiktok_ads/transform.js:115]:
```
let event = message.event?.toLowerCase();
if (!event) {
    throw new CustomError("Event name is required", 400);
  }
```

File: v0/destinations/tiktok_ads/transform.js [hunk 6]
```diff
@@ -110,9 +100,48 @@ const trackResponseBuilder = async (message, { Config }) => {
...
+
+  const standardEventsMap = getHashFromArrayWithDuplicate(eventsToStandard);
+  const trimmedEvent = event.toLowerCase().trim();
```
Comment (utsabc) [inline v0/destinations/tiktok_ads/transform.js:118]: This is redundant code we can do this when we are doing lower case

File: v0/destinations/tiktok_ads/transform.js [hunk 7]
```diff
@@ -110,9 +100,48 @@ const trackResponseBuilder = async (message, { Config }) => {
...
+      if (key === trimmedEvent) {
+        standardEventsMap[trimmedEvent].forEach(val => {
+          returnArray.push(getTrackResponse(message, Config, val));
+        });
+      }
+    });
+    return returnArray;
+  }
+
+  const response = defaultRequestConfig();
```
Comment (utsabc) [inline v0/destinations/tiktok_ads/transform.js:139]: response is unused variable is this correct?

File: v0/destinations/tiktok_ads/transform.js (hunk 4 above)
Comment (rohithkaza) [reply v0/destinations/tiktok_ads/transform.js]: Removed

File: v0/destinations/tiktok_ads/transform.js (hunk 5 above)
Comment (rohithkaza) [reply v0/destinations/tiktok_ads/transform.js]: Done

File: v0/destinations/tiktok_ads/transform.js (hunk 7 above)
Comment (rohithkaza) [reply v0/destinations/tiktok_ads/transform.js]: Yes, removed it

**Review by akashrpo — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**